File: rust/lib/src/services/availability_checker.rs
use std::{rc::Rc, thread::sleep, time::Duration};
use anyhow::{anyhow, Result};

use super::service::Service;

// Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
pub struct AvailabilityChecker {
    service_id: String,

    to_check: Rc<dyn Service>,
}

impl AvailabilityChecker {
    pub fn new(service_id: &str, to_check: Rc<dyn Service>) -> AvailabilityChecker {
        return AvailabilityChecker{
            service_id: service_id.to_owned(),
            to_check,
        };
    }

    // Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
    pub fn wait_for_startup(&self, time_between_polls: &Duration, max_num_retries: u32) -> Result<()> {
        for i in 0..max_num_retries {
            if self.to_check.is_available() {
                return Ok(());
            }

            // Don't wait if we're on the last iteration of the loop, since we'd be waiting unnecessarily
            if i < max_num_retries - 1 {
                sleep(time_between_polls.to_owned());
            }
        }
        return Err(anyhow!(
            "Service '{}' did not become available despite polling {} times with {}ms between polls",
            self.service_id,
            max_num_retries,
            time_between_polls.as_millis(),
        ));
    }
}

File: golang/lib/services/availability_checker.go
/*
 * Copyright (c) 2020 - present Kurtosis Technologies LLC.
 * All Rights Reserved.
 */

package services

import (
	"github.com/palantir/stacktrace"
	"time"
)

// Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
type AvailabilityChecker interface {
	WaitForStartup(timeBetweenPolls time.Duration, maxNumRetries int) error
}

type DefaultAvailabilityChecker struct {
	// ID of the service being monitored
	serviceId ServiceID

	// The service being monitored
	toCheck Service
}

func NewDefaultAvailabilityChecker(serviceId ServiceID, toCheck Service) *DefaultAvailabilityChecker {
	return &DefaultAvailabilityChecker{serviceId: serviceId, toCheck: toCheck}
}

// Docs available at https://docs.kurtosistech.com/kurtosis-libs/lib-documentation
func (checker DefaultAvailabilityChecker) WaitForStartup(timeBetweenPolls time.Duration, maxNumRetries int) error {
	for i := 0; i < maxNumRetries; i++ {
		if checker.toCheck.IsAvailable() {
			return nil
		}

		// Don't wait if we're on the last iteration of the loop, since we'd be waiting unnecessarily
		if i < maxNumRetries - 1 {
			time.Sleep(timeBetweenPolls)
		}
	}
	return stacktrace.NewError(
		"Service '%v' did not become available despite polling %v times with %v between polls",
		checker.serviceId,
		maxNumRetries,
		timeBetweenPolls)
}
